noxfile.py

```diff
@@ -0,0 +1,261 @@
+"""
+Nox is a command-line tool that automates testing in multiple Python environments, similar to tox.
+Unlike tox, Nox uses a standard Python file for configuration.
+
+To run all sessions defined in the noxfile.py file, run the following command:
+$ nox
+
+To run a specific session, run the following command:
+$ nox -s <session-name>
+
+To run a session with additional arguments, run the following command:
+$ nox -s <session-name> -- <additional-arguments>
+
+To list all available sessions, run the following command:
+$ nox -l
+"""
+
+import nox
+from nox.sessions import Session
+
+# default sessions (sorted alphabetically)
+nox.options.sessions = ["bandit", "black", "flake8", "isort", "mypy", "unit-tests"]
+
+
+def _install_requirements(session: Session) -> None:
+    session.install("-r", "requirements-extras.txt")
+
+
+@nox.session(reuse_venv=True)
+def bandit(session: Session) -> None:
+    """Run bandit on cachi2 directory"""
+    _install_requirements(session)
+    session.run(
+        "bandit",
+        "-c",
+        "pyproject.toml",
+        "-r",
+        "cachi2",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def black(session: Session) -> None:
+    """Run black on cachi2 and tests directories and noxfile.py"""
+    _install_requirements(session)
+    session.run(
+        "black",
+        "--check",
+        "--diff",
+        "cachi2",
+        "tests",
+        "noxfile.py",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def flake8(session: Session) -> None:
+    """Run flake8 on cachi2 and tests directories"""
+    _install_requirements(session)
+    # TODO: Move flake8-pyproject to requirements-extras.txt
+    session.install("flake8-pyproject")
+    session.run(
+        "flake8",
+        "cachi2",
+        "tests",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def isort(session: Session) -> None:
+    """Run isort on cachi2 and tests directories and noxfile.py"""
+    _install_requirements(session)
+    session.run(
+        "isort",
+        "--check",
+        "--diff",
+        "--color",
+        "cachi2",
+        "tests",
+        "noxfile.py",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def mypy(session: Session) -> None:
+    """Run mypy on cachi2 and tests directories"""
+    _install_requirements(session)
+    session.run(
+        "mypy",
+        "--install-types",
+        "--non-interactive",
+        "cachi2",
+        "tests",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(name="unit-tests", reuse_venv=True, python=["3.9", "3.10", "3.11", "3.12"])
```

Comment:
nitpick:
`["3.9", "3.10", "3.11", "3.12"]`

In order to avoid changing multiple files at the same time, since noxfile is a Python script we could introduce a simple helper that would parse the `requires-python` field from the `pyproject.toml` file and enumerate the version set from the specifier (based on `packaging.specifiers.SpecifierSet`), i.e. if we adjust the specifier to `>=3.9, <=3.12` such helper would yield this exact list. That way, any time we add a new supported version of Python (e.g. 3.13) or deprecate the lower boundary (i.e. 3.9 in this case), `noxfile` will likely not need an adjustment because of such update.

Reply:
Do you mean something like that ?
```
def _get_supported_python_versions() -> list[str]:
    """
    Read pyproject.toml and return a list of all supported Python versions
    based on the requires-python specification.
    """
    pyproject_path = Path("pyproject.toml")
    pyproject = tomli.loads(pyproject_path.read_text())
    requires_python = pyproject["project"]["requires-python"]

    specifiers = SpecifierSet(requires_python)
    min_minor = None
    max_minor = None

    for spec in specifiers:
        minor = int(re.match(r"3\.(\d+)", spec.version).group(1))

        if spec.operator in (">=", ">"):
            if min_minor is None or minor > min_minor:
                min_minor = minor
        elif spec.operator in ("<=", "<"):
            if max_minor is None or minor < max_minor:
                max_minor = minor

    return [f"3.{minor}" for minor in range(min_minor, max_minor + 1)]
```

Reply:
Using `int(re.match(r"3\.(\d+)", spec.version).group(1))` regexes for this purpose seems like a big overkill to me when packaging already provides a `Version` object that can be compared.

Also,
```
if spec.operator in (">=", ">"):
            if min_minor is None or minor > min_minor:
                min_minor = minor
```
is incorrect for e.g. `>=3.6, >3.7` because your `min_minor` would be 7 but due to the `>` operator it has to be 8, so you'd have to further special case that. We're already making a compromise here assuming only Python3 as the major, my idea suggestion crumbles like a house of cards with a hypothetical Python4 release in general, because then you cannot enumerate the versions without scraping Python's release page for the minor releases for each major to produce a correct version range, i.e. not ending up with something like Python 3.40 if hypothetically Python 4 is released and compatibility isn't such a problem like it were with Python2-3 (unlikely).
Therefore, my suggestion to this operator handling is to make a further compromise and **only assume** the `>=, <=` operators because we're in full control over our `pyproject.toml` and we only ever add/remove Python versions when we know our project would work with them, IOW we wouldn't skip a version of Python, that makes no sense from our POV. This alone will fix the problem I mentioned in the first sentence.

```
...
min_version = None
max_version = None
...
version = Version(spec.version)
...
return [f"3.{minor}" for minor in range(min_version.minor, max_version.minor + 1)]
```
should do the trick

Reply:
Could we parse the versions from `classifiers` ?

Reply:
I did consider using classifiers in my original thought process, but ruled them out for being more error prone. This made me think again and I think my original worries weren't justified since we're in full control of the file, so I think we could give it a try.